File: algorithms/200.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution
{
private:
    void dfs(vector< vector<char> > &grid, vector< vector<bool> > &visited, int x, int y)
    {
        visited[x][y] = true;
        int dx[] = {0, 0, 1, -1};
        int dy[] = {1, -1, 0, 0};
        for ( int i = 0; i < 4; ++i )
            if ( x+dx[i] >= 0 && x+dx[i] < (int)grid.size() &&
                 y+dy[i] >= 0 && y+dy[i] < (int)grid[0].size() &&
                 !visited[x+dx[i]][y+dy[i]] && grid[x+dx[i]][y+dy[i]] == '1' )
                dfs(grid, visited, x+dx[i], y+dy[i]);
    }
public:
    int numIslands(vector< vector<char> >& grid)
    {
        if ( grid.empty() )
            return 0;
        vector< vector<bool> > visited(grid.size(), vector<bool>(grid[0].size(), false));
        int result = 0;
        for ( unsigned int i = 0; i < grid.size(); ++i )
        {
            for ( unsigned int j = 0; j < grid[0].size(); ++j )
            {
                if ( !visited[i][j] && grid[i][j] == '1' )
                {
                    result += 1;
                    dfs(grid, visited, i, j);
                }
            }
        }
        return result;
    }
};

int main()
{
    return 0;
}

File: algorithms/python3/200.py
class Solution:
    def numIslands(self, grid):
        """
        :type grid: List[List[str]]
        :rtype: int
        """
        if len(grid) == 0 or len(grid[0]) == 0: return 0
        m, n = len(grid), len(grid[0])

        visitable = lambda x, y: grid[x][y] == '1'
        def dfs(x, y):
            grid[x][y] = '0'
            if x-1 >= 0 and visitable(x-1, y):
                dfs(x-1, y)
            if x+1 < m and visitable(x+1, y):
                dfs(x+1, y)
            if y-1 >= 0 and visitable(x, y-1):
                dfs(x, y-1)
            if y+1 < n and visitable(x, y+1):
                dfs(x, y+1)

        res = 0
        for i in range(m):
            for j in range(n):
                if visitable(i, j):
                    res += 1
                    dfs(i, j)
        return res
